File: examples/contrib/kenken2.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""KenKen puzzle in Google CP Solver.

http://en.wikipedia.org/wiki/KenKen
'''
KenKen or KEN-KEN is a style of arithmetic and logical puzzle sharing
several characteristics with sudoku. The name comes from Japanese and
is translated as 'square wisdom' or 'cleverness squared'.
...
The objective is to fill the grid in with the digits 1 through 6 such that:

  * Each row contains exactly one of each digit
  * Each column contains exactly one of each digit
  * Each bold-outlined group of cells is a cage containing digits which
    achieve the specified result using the specified mathematical operation:
      addition (+),
      subtraction (-),
      multiplication (x),
      and division (/).
      (Unlike in Killer sudoku, digits may repeat within a group.)

...
More complex KenKen problems are formed using the principles described
above but omitting the symbols +, -, x and /, thus leaving them as
yet another unknown to be determined.
'''


The solution is:

  5 6 3 4 1 2
  6 1 4 5 2 3
  4 5 2 3 6 1
  3 4 1 2 5 6
  2 3 6 1 4 5
  1 2 5 6 3 4



This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

import sys
from functools import reduce

from ortools.constraint_solver.python import constraint_solver as cp

#
# Ensure that the sum of the segments
# in cc == res
#


def calc(cc, x, res):

    solver = list(x.values())[0].solver

    if len(cc) == 2:

        # for two operands there may be
        # a lot of variants

        c00, c01 = cc[0]
        c10, c11 = cc[1]
        a = x[c00 - 1, c01 - 1]
        b = x[c10 - 1, c11 - 1]

        r1 = solver.add_is_equal_cst_var(a + b, res)
        r2 = solver.add_is_equal_cst_var(a * b, res)
        r3 = solver.add_is_equal_var(a * res, b)
        r4 = solver.add_is_equal_var(b * res, a)
        r5 = solver.add_is_equal_cst_var(a - b, res)
        r6 = solver.add_is_equal_cst_var(b - a, res)
        solver.add(r1 + r2 + r3 + r4 + r5 + r6 >= 1)

    else:

        # res is either sum or product of the segment

        xx = [x[i[0] - 1, i[1] - 1] for i in cc]

        # Sum
        # # SumEquality don't work:
        # this_sum = solver.add_sum_equality(xx, res)
        this_sum = solver.add_is_equal_cst_var(solver.sum(xx), res)

        # Product
        # # Prod (or MakeProd) don't work:
        # this_prod = solver.add_is_equal_cst_var(solver.Prod(xx), res)
        this_prod = solver.add_is_equal_cst_var(reduce(lambda a, b: a * b, xx), res)
        solver.add(this_sum + this_prod >= 1)


def main():

    # Create the solver.
    solver = cp.Solver("KenKen")

    #
    # data
    #

    # size of matrix
    n = 6

    # For a better view of the problem, see
    #  http://en.wikipedia.org/wiki/File:KenKenProblem.svg

    # hints
    #    [sum, [segments]]
    # Note: 1-based
    problem = [
        [11, [[1, 1], [2, 1]]],
        [2, [[1, 2], [1, 3]]],
        [20, [[1, 4], [2, 4]]],
        [6, [[1, 5], [1, 6], [2, 6], [3, 6]]],
        [3, [[2, 2], [2, 3]]],
        [3, [[2, 5], [3, 5]]],
        [240, [[3, 1], [3, 2], [4, 1], [4, 2]]],
        [6, [[3, 3], [3, 4]]],
        [6, [[4, 3], [5, 3]]],
        [7, [[4, 4], [5, 4], [5, 5]]],
        [30, [[4, 5], [4, 6]]],
        [6, [[5, 1], [5, 2]]],
        [9, [[5, 6], [6, 6]]],
        [8, [[6, 1], [6, 2], [6, 3]]],
        [2, [[6, 4], [6, 5]]],
    ]

    num_p = len(problem)

    #
    # variables
    #

    # the set
    x = {}
    for i in range(n):
        for j in range(n):
            x[i, j] = solver.new_int_var(1, n, "x[%i,%i]" % (i, j))

    x_flat = [x[i, j] for i in range(n) for j in range(n)]

    #
    # constraints
    #

    # all rows and columns must be unique
    for i in range(n):
        row = [x[i, j] for j in range(n)]
        solver.add_all_different(row)

        col = [x[j, i] for j in range(n)]
        solver.add_all_different(col)

    # calculate the segments
    for res, segment in problem:
        calc(segment, x, res)

    #
    # search and solution
    #
    db = solver.phase(
        x_flat,
        cp.IntVarStrategy.INT_VAR_DEFAULT,
        cp.IntValueStrategy.INT_VALUE_DEFAULT,
    )

    solver.new_search(db)

    num_solutions = 0
    while solver.next_solution():
        for i in range(n):
            for j in range(n):
                print(x[i, j].value(), end=" ")
            print()

        print()
        num_solutions += 1

    solver.end_search()

    print()
    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


if __name__ == "__main__":
    main()

File: examples/contrib/kenken2.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Google.OrTools.ConstraintSolver;

public class KenKen2
{
    /**
     * Ensure that the sum of the segments
     * in cc == res
     *
     */
    public static void calc(Solver solver, int[] cc, IntVar[,] x, int res)
    {
        int ccLen = cc.Length;
        if (ccLen == 4)
        {
            // for two operands there's
            // a lot of possible variants
            IntVar a = x[cc[0] - 1, cc[1] - 1];
            IntVar b = x[cc[2] - 1, cc[3] - 1];

            IntVar r1 = a + b == res;
            IntVar r2 = a * b == res;
            IntVar r3 = a * res == b;
            IntVar r4 = b * res == a;
            IntVar r5 = a - b == res;
            IntVar r6 = b - a == res;

            solver.Add(r1 + r2 + r3 + r4 + r5 + r6 >= 1);
        }
        else
        {
            // For length > 2 then res is either the sum
            // the the product of the segment

            // sum the numbers
            int len = cc.Length / 2;
            IntVar[] xx = (from i in Enumerable.Range(0, len) select x[cc[i * 2] - 1, cc[i * 2 + 1] - 1]).ToArray();

            // Sum
            IntVar this_sum = xx.Sum() == res;

            // Product
            // IntVar this_prod = (xx.Prod() == res).Var(); // don't work
            IntVar this_prod;
            if (xx.Length == 3)
            {
                this_prod = (x[cc[0] - 1, cc[1] - 1] * x[cc[2] - 1, cc[3] - 1] * x[cc[4] - 1, cc[5] - 1]) == res;
            }
            else
            {
                this_prod = (x[cc[0] - 1, cc[1] - 1] * x[cc[2] - 1, cc[3] - 1] * x[cc[4] - 1, cc[5] - 1] *
                             x[cc[6] - 1, cc[7] - 1]) == res;
            }

            solver.Add(this_sum + this_prod >= 1);
        }
    }

    /**
     *
     * KenKen puzzle.
     *
     * http://en.wikipedia.org/wiki/KenKen
     * """
     * KenKen or KEN-KEN is a style of arithmetic and logical puzzle sharing
     * several characteristics with sudoku. The name comes from Japanese and
     * is translated as 'square wisdom' or 'cleverness squared'.
     * ...
     * The objective is to fill the grid in with the digits 1 through 6 such that:
     *
     * * Each row contains exactly one of each digit
     * * Each column contains exactly one of each digit
     * * Each bold-outlined group of cells is a cage containing digits which
     *   achieve the specified result using the specified mathematical operation:
     *     addition (+),
     *     subtraction (-),
     *     multiplication (x),
     *     and division (/).
     *    (Unlike in Killer sudoku, digits may repeat within a group.)
     *
     * ...
     * More complex KenKen problems are formed using the principles described
     * above but omitting the symbols +, -, x and /, thus leaving them as
     * yet another unknown to be determined.
     * """
     *
     * The solution is:
     *
     *    5 6 3 4 1 2
     *    6 1 4 5 2 3
     *    4 5 2 3 6 1
     *    3 4 1 2 5 6
     *    2 3 6 1 4 5
     *    1 2 5 6 3 4
     *
     *
     * Also see http://www.hakank.org/or-tools/kenken2.py
     * though this C# model has another representation of
     * the problem instance.
     *
     */
    private static void Solve()
    {
        Solver solver = new Solver("KenKen2");

        // size of matrix
        int n = 6;
        IEnumerable<int> RANGE = Enumerable.Range(0, n);

        // For a better view of the problem, see
        //  http://en.wikipedia.org/wiki/File:KenKenProblem.svg

        // hints
        //  sum, the hints
        // Note: this is 1-based
        int[][] problem = { new int[] { 11, 1, 1, 2, 1 },
                            new int[] { 2, 1, 2, 1, 3 },
                            new int[] { 20, 1, 4, 2, 4 },
                            new int[] { 6, 1, 5, 1, 6, 2, 6, 3, 6 },
                            new int[] { 3, 2, 2, 2, 3 },
                            new int[] { 3, 2, 5, 3, 5 },
                            new int[] { 240, 3, 1, 3, 2, 4, 1, 4, 2 },
                            new int[] { 6, 3, 3, 3, 4 },
                            new int[] { 6, 4, 3, 5, 3 },
                            new int[] { 7, 4, 4, 5, 4, 5, 5 },
                            new int[] { 30, 4, 5, 4, 6 },
                            new int[] { 6, 5, 1, 5, 2 },
                            new int[] { 9, 5, 6, 6, 6 },
                            new int[] { 8, 6, 1, 6, 2, 6, 3 },
                            new int[] { 2, 6, 4, 6, 5 } };

        int num_p = problem.GetLength(0); // Number of segments

        //
        // Decision variables
        //
        IntVar[,] x = solver.MakeIntVarMatrix(n, n, 1, n, "x");
        IntVar[] x_flat = x.Flatten();

        //
        // Constraints
        //

        //
        //  alldifferent rows and columns
        foreach (int i in RANGE)
        {
            // rows
            solver.Add((from j in RANGE select x[i, j]).ToArray().AllDifferent());

            // cols
            solver.Add((from j in RANGE select x[j, i]).ToArray().AllDifferent());
        }

        // Calculate the segments
        for (int i = 0; i < num_p; i++)
        {
            int[] segment = problem[i];

            // Remove the sum from the segment
            int len = segment.Length - 1;
            int[] s2 = new int[len];
            Array.Copy(segment, 1, s2, 0, len);

            // sum this segment
            calc(solver, s2, x, segment[0]);
        }

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(x_flat, Solver.INT_VAR_DEFAULT, Solver.INT_VALUE_DEFAULT);

        solver.NewSearch(db);

        while (solver.NextSolution())
        {
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    Console.Write(x[i, j].Value() + " ");
                }
                Console.WriteLine();
            }
            Console.WriteLine();
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        Solve();
    }
}
